### ISSUE TITLE
(Profiled) Code spends most of its time in `Internal.Devel.check`

### ISSUE THREAD

**Issue by fhaust** [haskell-numerics/hmatrix#317]
I just profiled some `hmatrix` heavy code and found that it spends 70% of its time in `Internal.Devel.check` which sounds strange to me. From what I gathered that function is used to check the return code of the FPU?

To be fair, this might be an non issue when the code is running not profiled due to optimizations kicking in?

**Comment by idontgetoutmuch** [comment]
I have no idea - if you could post an example, I could investigate a bit further.

**Comment by StefanHubner** [comment]
I just found the same. I attached the profile file. 
Indeed, this could just be an FFI call being attributed to that function by the profiler.

[profile.zip](https://github.com/haskell-numerics/hmatrix/files/4707769/profile.zip)

**Comment by idontgetoutmuch** [comment]
@StefanHubner Thanks for investigating - just to be clear: this speed up only happens when profiling? So the problem seems to be one with profiling not with hmatrix itself?

**Comment by StefanHubner** [comment]
No, the speed-up came from it never being evaluated. The error code checking is not costly but the computational expense of the FFI call is attributed to the check function. 
I hope this makes sense!

**Comment by dschrempf** [comment]
I am having the same issue. It seems that the FFI call is evaluated when the error value is checked, and so, the computation happens in `Internal.Deve.check`. Is it possible to change this behavior to get more informative profiling results?

**Comment by HuwCampbell** [comment]
I think we could very well inline calls to `check` and `(#|)`, their definitions are quite small, and it would probably push the cost centre to the c calls. There's also the option of explicit cost centre annotations.
